### ISSUE TITLE
Behavior when offline

### ISSUE THREAD

**Issue by simonemarin** [signalpoint/DrupalGap#146]
Assuming our users are not always online it would be desirable for the app to use the cached content when offline after having notified the user that the content may be obsolete and that CRUD operation are available only on a partial read mode.
Simone

**Comment by signalpoint** [comment]
Agreed. The DG System Connect resource success callback would need to be adjusted to cache the 'field info instances' so any subsequent loads would have that information (https://github.com/signalpoint/DrupalGap/blob/7.x-1.x-alpha/modules/services/drupalgap_system.js).

I think the offline page could use some improvement for a typical use case (https://github.com/signalpoint/DrupalGap/blob/7.x-1.x-alpha/modules/system/system.js). I also think the Getting Started guide should explain how to overwrite the page callback for a certain path, so we design our own 'offline' page and functionality.

Something similar to hook_menu_alter() in Drupal would be ideal for developers wishing to overwrite page callback functions defined by hook_menu() in DG.

**Comment by scotthooker** [comment]
+1 for a hook_menu_alter

Very important we have some way of telling the user that they aren't connected. E.g. I have a pull to refresh functionality that in the new version just hangs as there is no internet connection.

Need some way on a service call to return a message letting the user know they aren't connected.

**Comment by signalpoint** [comment]
Although hook_menu_alter() doesn't exist yet, this can be accomplished within hook_deviceready(). For example we can change the page_callback for the system's 'dashboard' page:

```
/**
 * Implements hook_deviceready().
 */
function my_module_deviceready() {
  try {
    //dpm(drupalgap.menu_links);
    drupalgap.menu_links['dashboard'].page_callback = 'my_module_dashboard_page';
  }
  catch (error) { console.log('my_module_deviceready - ' + error); }
}

function my_module_dashboard_page() {
  try {
    var content = {};
    content.into = {
      markup: '<p>Clean your dashboard!</p>'
    };
    return content;
  }
  catch (error) { console.log('my_module_dashboard_page - ' + error); }
}
```

Just use `dpm(drupalgap.menu_links);` to look for the particular menu path that you'd like to modify.

I've updated the documentation to include an example of altering an existing menu link:

http://drupalgap.org/node/210

This issue does still stand though, there needs to be a better way to handle when the device is offline, or if it loses connection.

**Comment by signalpoint** [comment]
This issue has been closed in favor of this issue: #258
